library/std/src/sys/unix/path/posix_path.rs

```diff
@@ -0,0 +1,50 @@
+use crate::ffi::{CStr, OsString};
+use crate::io;
+use crate::path::{Path, PathBuf};
+use crate::sys::common::small_c_string::run_path_with_cstr;
+use crate::sys::os_str::Buf;
+use crate::sys_common::FromInner;
+
+pub type NativePath = CStr;
+
+#[unstable(feature = "fs_native_path_internals", issue = "none")]
+pub trait PathLike: crate::sealed::Sealed {
+    fn with_path<T, F: FnOnce(&Path) -> io::Result<T>>(self, f: F) -> io::Result<T>;
+    fn with_native_path<T, F: FnOnce(&NativePath) -> io::Result<T>>(self, f: F) -> io::Result<T>;
+}
+
+#[unstable(feature = "fs_native_path_internals", issue = "none")]
+impl<P: AsRef<Path>> PathLike for P {
+    fn with_path<T, F: FnOnce(&Path) -> io::Result<T>>(self, f: F) -> io::Result<T> {
+        f(self.as_ref())
+    }
+    fn with_native_path<T, F: FnOnce(&CStr) -> io::Result<T>>(self, f: F) -> io::Result<T> {
+        run_path_with_cstr(self.as_ref(), f)
+    }
+}
+
+#[unstable(feature = "fs_native_path_internals", issue = "none")]
+impl PathLike for &CStr {
+    fn with_path<T, F: FnOnce(&Path) -> io::Result<T>>(self, f: F) -> io::Result<T> {
+        let osstr: OsString = FromInner::from_inner(Buf { inner: self.to_bytes().to_vec() });
+        let path = PathBuf::from(osstr);
+        f(&path)
```

Comment:
Why the allocation here?
```suggestion
        let osstr = OsStr::from_bytes(self.to_bytes());
        f(osstr.as_ref())
```

Reply:
Ah, good point. This was left over from a refactoring.